### ISSUE TITLE
Approach to model selection?

### ISSUE THREAD

**Issue by tbuehrens** [casruff/Skagit-River-Steelhead-Forecast#3]
Hey @casruff, your past approach of calculating model error based on a series of 1-year ahead forecasts was rock solid and is exactly what Mark and I do in another manuscript we're working on. LOOIC is also a defensible fall-back for comparing AR1 with AR1MA1, since the AR1MA1 model was taking 24 hrs to run. I have a couple thoughts though:

1) Models *should* be alot faster now that I parallelized them (~75% reduction in run time)
2) That still doesn't get us to a point where re-running each 5 or 10 times is "fast".
3. I forgot the other day that the stan group who develop LOOIC also developed "Approximate leave-future-out cross-validation for Bayesian time series models" here: https://mc-stan.org/loo/articles/loo2-lfo.html. Maybe worth comparing? I'd guess it will give us the same results. Do you want me to test this?

**Comment by casruff** [comment]
Hey Thomas,

I would be interested in seeing the the results of the approximate leave-future-cross-validation methods if it isn't too much trouble. Thanks,

Casey

**Comment by tbuehrens** [comment]
1. As you see from the code you pulled from my fork I started the approx leave future out CV but haven't yet satisfied myself it is working for us..
2. For your one step ahead loops, I noticed you are currently using the jags_fit function rather than jags_fit2. I'm currently pulling your updates to your test branch to mine but it's taking a long time due to big files...once I'm done with that I'll try to implement a commit making this change to allow for parallelizing the one-step-ahead forecasts. Sound good?

**Comment by casruff** [comment]
Cool. Whoops, I forgot to double check that I was using the new jags_fit2 function. Thank you for catching that.  Also, if you  have any suggestions on improving the one step ahead loops (i.e. making them less clunky), I'm definitely open to them.

**Comment by casruff** [comment]
Sorry for the big files..33 saved jags model objects.

**Comment by tbuehrens** [comment]
ok, just finished the pulll...why are the AR1MA1 files so much bigger? Do they have the same  MCMC control params or do they have more posterior samples?

**Comment by casruff** [comment]
Good catch. I first ran the the AR1MA1 models with longer chain length and burn. After the first 11 models were saved and the loop moved onto the AR1 model, the code threw an error because I had the data index counter (c) in the wrong place. So, I fixed the problem and then bumped down the MCMC params for the sets of AR1 and AR1_resid models/forecasts that needed to still fit and run for the purposes of making sure the code run. Sorry for the confusion.

**Comment by tbuehrens** [comment]
no worries, are you going to re-run the AR1 and AR1_resid with same iterations as AR1MA1? I think we'd be almost at the finish line then! (or do you want me to?)

**Comment by tbuehrens** [comment]
@casruff I made a little progress on the approximate LFO. Here is a graph of the delta-LFO-IC (like LOOIC) using 20 years as a training set and the last 12 years as one-step-ahead forecasts for the LFO calculation:
![image](https://user-images.githubusercontent.com/40445951/102560630-8c16d000-4087-11eb-82ce-8a46026f5d62.png)
the weights are:
![image](https://user-images.githubusercontent.com/40445951/102560659-a18bfa00-4087-11eb-81bc-2eb20da7cbc4.png)
This suggests the MA1_AR1 is best with the AR1_resid second best and little weight to the AR1. However, many of the pareto K values were high, which I was surprised by because the pareto K values as estimated by LOOIC were very good. I asked a question on the Stan Forum to the guy who wrote the paper to see if we can figure out whether the pareto K results are an indication I did something wrong, or expected. 

You making any progress on the one step ahead code?

**Comment by tbuehrens** [comment]
I should also note that the results appear to be quite sensitive to a) whether I only include escapement data in the LFO-IC, or whether I include the age data too, and b) how many years I compute LFO for....though the Pareto scores get worse the more years I calculate it for.

EDIT: good blog post by Aki Vehtari (a badass) on use of LOO (and thus LOOIC) for evaluating timeseries...in short, he says not great for evaluating prediction skill.
https://statmodeling.stat.columbia.edu/2018/08/03/loo-cross-validation-approaches-valid/

**Comment by tbuehrens** [comment]
@casruff 
Woot Woot!!! I think i'm cooking with gas...I used your re-fits to calculate approx LFO-CV when the Pareto K was too high...now, results are consistent regardless of whether I perform with just escapement data or escapement plus age. Here are results:
![image](https://user-images.githubusercontent.com/40445951/102642274-4b17cd80-4112-11eb-8359-913f7328c18c.png)
![image](https://user-images.githubusercontent.com/40445951/102642307-523edb80-4112-11eb-9e85-e386588a2289.png)
After re-running the one-step-ahead forecasts so they have slightly more iterations, I believe we now have a solution for model averaging! (table is delta-LFO-IC as well as model weights (w)...obviously exact values may change when we have better convergence. Also, still think we may want to plot one-step-ahead fits of all three models to convince folks they work and allow visualization of comparison of fits...will push this code to my test branch of my fork

**Comment by casruff** [comment]
Hey Thomas,

Sorry for the radio silence. This is great stuff. So did you rerun the 1 step ahead forecasts with more iterations? I'm currently rerunning them on my machine with more iterations and they should be done later today. I started working on a new model selection table that includes metrics such as RMSE but it seems like we don't need that now with the LFO-IC working now?

Casey

**Comment by tbuehrens** [comment]
@casruff  I didn't rerun them...if you get the rmse/nape/mase/mlar table set up, then we can compare those with LFO-CV. I'd also like to plot the fits...we're nearly nearly there. Once your code is done, I can integrate my edits into yours...

**Comment by tbuehrens** [comment]
@casruff Hey casey, can you push your re-fits of the model with higher iterations to your test branch? I'd like to re-run my code with them to see if results match with better convergence

**Comment by casruff** [comment]
Hey Thomas,

I pushed the  re-fits of the one step ahead forecasts with higher iterations to the test branch. It looks like model selection via RMSE (comparing forecasts to observed terminal run size) and LOOIC based on the elpd of the models fit to the most recent set of years through RY 2020 both indicate that the AR1 model fits the data best, followed by AR1_resid and AR1MA1. I haven't updated the results section to include plots of the one step ahead forecasts to the observations. 

Casey

**Comment by casruff** [comment]
Fore some reason my computer froze when I went to push so it  looks like the updates weren't successfully updated on my test branch. Updating now.

**Comment by casruff** [comment]
test branch is now updated with the new model fits.

**Comment by tbuehrens** [comment]
@casruff cool thanks. I've working on configuring some software on my laptop and then I'll merge in the changes from your repo and then run the LFO-CV code...guessing it will provide us a means to accomplish model-averaging that will be quite robust

**Comment by tbuehrens** [comment]
Ok: 
1. the good news, approx LFO-CV agrees loosely with what you wrote above (AR1 best, AR1MA1 best for LFO-CV, AR1_resid 3rd best), when I use the 10 leave out one years you did. 
![image](https://user-images.githubusercontent.com/40445951/102831584-048bd280-43a1-11eb-9e1a-b01fd9570f18.png)
2. With these new model fits, that result remains largely unchanged when I include 23 years of approx-LFO-CV (starting fitting with a min of 20 years of data). 
3. It also remains unchanged when I excluded or include age data (while always including escapement).
4. Result is also unchanged when I use original full fits in combination with the new refits
5. Really my only hesitation is verifying reasonable convergence was achieved...we could write some code to loop through producing some summary stats on number of effective draws and I'd feel better, but I'm guessing we are just about there. Thanks for re-fitting these!

This is promising, and implies that these new refits have switched weight from the MA1AR1 over to the AR (original) pretty convincingly. Further supported by the RMSE , MAPE, MSE results....

We have a few different sets of model weights we can use: 1) RMSE weights, 2, LOOIC bma weights, 3) LFO-CV bma weights....i'd tend to advocate for 1 or 3...maybe we can find a time to chat briefly tomorrow or weds???

Final note:  I pulled your 'test' repo and merged it with mine...currently pushing my updates to my 'test' repo. You can then pull request them into your test when my push has completed (still working at 4 pm right now). You'll know my 'test' is done when the "saved_models" folder disappears in my test repo (i pulled those results out and pasted them in the main cache folder).

**Comment by tbuehrens** [comment]
@casruff OK, got the model code tweaked and re-ran the AR_resid and  AR1MA1 super short runs with 1000 burn in 1000 kept, 4 chains and no thinning. Need to re-run longer. I also set up the model selection code so the LFO-IC is joined with your table and we have multiple weights available to select from. Finally I also made this plot of the 10 years worth of one-step-ahead forecasts and the observed catch+escapement:
![image](https://user-images.githubusercontent.com/40445951/102944712-6b2bf180-4470-11eb-910d-be478b9b1a0c.png)
here is teh model selection table...much more even weights
![image](https://user-images.githubusercontent.com/40445951/102944768-9b739000-4470-11eb-964b-201721c2767c.png)
...except for LFO-IC (and I'm sure LOOIC too) both of which are not reliable right now because the fits to the different  models contain different numbers of draws and convergence/lack thereof...when we re-run the two models with the same iterations as your original runs, the LFO-IC should be very useful to.
